pkg/csi/controller/server.go

```diff
@@ -272,6 +277,79 @@ func (c *Server) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
 	return &csi.DeleteVolumeResponse{}, nil
 }
 
+// ControllerExpandVolume - controller RPC to expand volume
+// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume
+func (c *Server) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
+	klog.V(3).InfoS("Expand volume requested", "name", req.GetVolumeId())
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty volume ID in the request")
+	}
+	if req.CapacityRange == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid capacity range in the request for volume %v expansion", volumeID)
+	}
+	requiredBytes := req.CapacityRange.RequiredBytes
+
+	volume, err := client.VolumeClient().Get(ctx, volumeID, metav1.GetOptions{
+		TypeMeta: types.NewVolumeTypeMeta(),
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get volume %v; %v", volumeID, err)
+	}
+
+	if volume.Status.TotalCapacity >= requiredBytes {
+		// As per the specification, nothing to do
+		return &csi.ControllerExpandVolumeResponse{CapacityBytes: requiredBytes}, nil
+	}
+
+	drive, err := client.DriveClient().Get(
+		ctx, string(volume.GetDriveID()), metav1.GetOptions{TypeMeta: types.NewDriveTypeMeta()},
+	)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"unable to get drive %v for volume %v; %v",
+			volume.GetDriveID(), volumeID, err,
+		)
+	}
+	size := requiredBytes - volume.Status.TotalCapacity
+	if size > drive.Status.FreeCapacity {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"required bytes %v is greater than free capacity of drive %v for volume %v expansion",
+			requiredBytes, volume.GetDriveID(), volumeID,
+		)
+	}
+	drive.Status.FreeCapacity -= size
+	drive.Status.AllocatedCapacity += size
+
+	volume.Status.TotalCapacity = requiredBytes
+	volume.Status.AvailableCapacity = volume.Status.TotalCapacity - volume.Status.UsedCapacity
+	_, err = client.VolumeClient().Update(ctx, volume, metav1.UpdateOptions{
+		TypeMeta: types.NewVolumeTypeMeta(),
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to update volume %v; %v", volumeID, err)
+	}
+
+	_, err = client.DriveClient().Update(
+		ctx, drive, metav1.UpdateOptions{TypeMeta: types.NewDriveTypeMeta()},
+	)
+	if err != nil {
```

Comment:
when a drive update fails here due to some reason, it never gets updated again as the volume got updated already with the expanded size and we have the following entry check. Due to this, the corresponding drive will not be updated and would still have older values for sizes.

```go
if volume.Status.TotalCapacity >= requiredBytes {
	// As per the specification, nothing to do
	return &csi.ControllerExpandVolumeResponse{CapacityBytes: requiredBytes}, nil
}
```

To make it idempotent, I think we can add a field to the directpvvolume object to remember the expansion state. We can refer this state to know if the corresponding drive is updated or not..

Reply:
This idea wouldn't work because it is racy.
1. In directpvvolume, we need to have the field set to false before drive update
2. after successful drive update, if we fail to set the field to true, it is hard to know what happened lastly.

The solution could be drive controller should update allocated/free by calculating values from volumes periodically.